### PR TITLE
OCPBUGS-46035: fix skew support for node-joiner

### PR DESCRIPTION
**Pull request by andfasano**
This patch allows to run the `oc adm node-image` commands using oc 4.17.x against a cluster 4.18

### CODE REVIEW

**Review by rwsu — COMMENTED** [review]
/lgtm

File: pkg/asset/agent/joiner/clusterinfo.go [hunk 1]
```diff
@@ -254,6 +254,16 @@ func (ci *ClusterInfo) retrieveArchitecture() error {
...
 	workerMachineConfig, err := ci.OpenshiftMachineConfigClient.MachineconfigurationV1().MachineConfigs().Get(context.Background(), "99-worker-fips", metav1.GetOptions{})
 	if err != nil {
+		// Older oc client may not have sufficient permissions,
+		// falling back to previous implementation.
+		if errors.IsForbidden(err) {
+			installConfig, err := ci.retrieveInstallConfig()
+			if err != nil {
+				return err
+			}
+			ci.FIPS = installConfig.FIPS
```
Comment (bfournie) [inline pkg/asset/agent/joiner/clusterinfo.go:264]: retrieveInstallConfig can return `nil, nil` at line 351.
Seems that you need to check installConfig != nil before using it

File: pkg/asset/agent/joiner/clusterinfo.go [hunk 2]
```diff
@@ -346,6 +401,16 @@ func (ci *ClusterInfo) retrieveImageDigestMirrorSets() error {
...
 	imageContentPolicies, err := ci.OpenshiftClient.ConfigV1().ImageContentPolicies().List(context.Background(), metav1.ListOptions{})
 	if err != nil {
+		// Older oc client may not have sufficient permissions,
+		// falling back to previous implementation.
+		if errors.IsForbidden(err) {
+			installConfig, err := ci.retrieveInstallConfig()
+			if err != nil {
+				return err
+			}
+			ci.DeprecatedImageContentSources = installConfig.DeprecatedImageContentSources
```
Comment (bfournie) [inline pkg/asset/agent/joiner/clusterinfo.go:411]: Same comment about checking installConfig

File: pkg/asset/agent/joiner/clusterinfo.go [hunk 3]
```diff
@@ -273,6 +283,16 @@ func (ci *ClusterInfo) retrieveSSHKey() error {
...
 	workerMachineConfig, err := ci.OpenshiftMachineConfigClient.MachineconfigurationV1().MachineConfigs().Get(context.Background(), "99-worker-ssh", metav1.GetOptions{})
 	if err != nil {
+		// Older oc client may not have sufficient permissions,
+		// falling back to previous implementation.
+		if errors.IsForbidden(err) {
+			installConfig, err := ci.retrieveInstallConfig()
+			if err != nil {
+				return err
+			}
+			ci.SSHKey = installConfig.SSHKey
```
Comment (bfournie) [inline pkg/asset/agent/joiner/clusterinfo.go:293]: Same comment about checking installConfig != nil

File: pkg/asset/agent/joiner/clusterinfo.go [hunk 4]
```diff
@@ -319,9 +344,39 @@ func (ci *ClusterInfo) retrieveBootArtifactsBaseURL() error {
 	return nil
 }
 
+func (ci *ClusterInfo) retrieveInstallConfig() (*types.InstallConfig, error) {
+	clusterConfig, err := ci.Client.CoreV1().ConfigMaps("kube-system").Get(context.Background(), "cluster-config-v1", metav1.GetOptions{})
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return nil, nil
```
Comment (bfournie) [inline pkg/asset/agent/joiner/clusterinfo.go:351]: Maybe need a unit test for this?

File: pkg/asset/agent/joiner/clusterinfo.go (hunk 1 above)
Comment (andfasano) [reply pkg/asset/agent/joiner/clusterinfo.go:264]: Nice catch, thanks @bfournie !
Fixed as per previous implementation

**Comment by andfasano** [comment]
A backport to 4.18 will be required by https://github.com/openshift/installer/pull/9265

**Review by pawanpinjarkar — APPROVED** [review]
/approve

**Comment by andfasano** [comment]
/override e2e-azure-ovn-upi

Since there is an e2e failure not related to the current patch

**Pull request merged**